tests/variance_test_1.yaml

```diff
@@ -0,0 +1,26 @@
+---
+program: variance
+inputs:
+  array_0:
+    SecretInteger: '5'
+  array_1:
+    SecretInteger: '3'
+  array_2:
+    SecretInteger: '1'
+  array_3:
+    SecretInteger: '12'
+  array_4:
+    SecretInteger: '3'
+  array_5:
+    SecretInteger: '1'
+  array_6:
+    SecretInteger: '5'
+  array_7:
+    SecretInteger: '3'
+  array_8:
+    SecretInteger: '1'
+  array_9:
+    SecretInteger: '70'
+expected_outputs:
+  variance:
+    SecretInteger: '404'
```

Comment:
shouldn't this be 449?

Reply:
Thanks for your feedback. I believe the results you were expecting were for a sample dataset, where the variance is calculated by dividing the sum of squares by $n-1$. In this version of the program, we defined the variance for the population, which is defined by dividing the sum of squares by $n$.
You can check that [here](https://www.calculatorsoup.com/calculators/statistics/variance-calculator.php).

I assume the following comments are also about the sample variance, so I am skipping them.

Reply:
- [x] `toml` file has been updated.
- [x] Files has been renamed to `variance_integer`. 

I am not sure about how to perform the computations over fixed point numbers. I check the examples in the link you provided, where rational numbers can be defined like this `c = na.rational(1.2)`. Could you provide an example about how to read a Nada-Numpy array as a `secret_rational` instead of `SecretInteger`.

Reply:
https://github.com/NillionNetwork/nada-numpy/blob/main/examples/dot_product/src/dot_product.py

If you have a look at this example and write na.SecretRational where it says SecretInteger.

Reply:
I have the `variance_fp.py` file as follows:

```
from nada_dsl import *
import nada_numpy as na
from nada_numpy import SecretRational, Rational

DIM = 10

def mean(array: na.NadaArray):
    sum = Rational(Integer(0))

    # Compute the sum
    for v in array:
        sum += v

    # Divide the sum by the number of the elements
    m = sum / Rational(Integer(DIM))
    return m

def variance(array: na.NadaArray):
    # Compute the mean
    m = mean(array)

    # Initialize the sum to 0
    sum = Rational(Integer(0))

    # Compute (x_i - mean)^2
    for v in array:
        diff = v - m
        sum += (diff * diff)

    var = sum / Rational(Integer(DIM))
    return  var

def nada_main():
    party1 = Party(name="Party1")

    arr = na.array([DIM], party1, "array", SecretRational)

    var = variance(arr)

    return na.output(var, party1, "variance")
```

And the test file `variance_fp_test_1.yaml` as follows:

```
---
program: variance_fp
inputs:
  array_0:
    SecretInteger: '5'
  array_1:
    SecretInteger: '3'
  array_2:
    SecretInteger: '1'
  array_3:
    SecretInteger: '12'
  array_4:
    SecretInteger: '3'
  array_5:
    SecretInteger: '1'
  array_6:
    SecretInteger: '5'
  array_7:
    SecretInteger: '3'
  array_8:
    SecretInteger: '1'
  array_9:
    SecretInteger: '70'
expected_outputs:
  variance:
    SecretInteger: '404'
```

It gives an overflowed result:
```
Outputs: {
    "variance": SecretInteger(
        NadaInt(
            464528913203,
        ),
    ),
}
```

I tried to change the `SecretInteger` datatype in yaml to `SecretRational`, but it gives an error which indicates that `SecretRational` is not known.

```
0: expected_outputs: unknown variant `SecretRational`, expected one of `Integer`, `UnsignedInteger`, `Boolean`, `SecretInteger`, `SecretUnsignedInteger`, `SecretNonZeroInteger`, `SecretNonZeroUnsignedInteger`, `SecretBoolean`, `SecretBlob`, `ShamirShareInteger`, `ShamirShareUnsignedInteger`, `ShamirShareBoolean`, `ShamirParticleInteger`, `ShamirParticleUnsignedInteger`, `ShamirParticleBoolean`, `Array`, `Tuple` at line 25 column 11
```

Reply:
na.SecretRational I think. 

@mathias-nillion can help further as he wrote some of that library. :)

Reply:
@OmarAlmighty Oh right, indeed `np.var(np.array([5,3,1,12,3,1,5,3,1,70]))` returns the correct result!